File: backend/internal/recipe/recipe.go
// Package recipe holds the Go side of the Go-to-Dart recipe contract
// (schema/recipe.schema.json, ADR 0002, ADR 0004): the Draft type the service
// returns and the structural validator the post-process stage runs on every
// Draft the model produces.
//
// A Draft holds nothing about the Generation Request that produced it (H16).
package recipe

import (
	_ "embed"
	"time"
)

// SchemaJSON is a byte-identical copy of schema/recipe.schema.json, embedded so
// the binary carries its own contract. A test asserts it has not drifted.
//
//go:embed recipe.schema.json
var SchemaJSON []byte

// SchemaVersion is the value of Draft.schemaVersion this service writes.
const SchemaVersion = 1

// Units is the closed, metric-canonical unit set (ADR 0004).
var Units = []string{"g", "ml", "tsp", "tbsp", "piece", "clove", "slice", "sprig", "bunch", "pinch"}

// MealTypes is the closed Meal Type set.
var MealTypes = []string{"breakfast", "lunch", "dinner", "snack", "dessert", "drink"}

// Draft is what the Go service returns: a Body plus identity and provenance.
type Draft struct {
	ID            string    `json:"id"`
	SchemaVersion int       `json:"schemaVersion"`
	Source        Source    `json:"source"`
	CreatedAt     time.Time `json:"createdAt"`
	Body
}

// Body is the recipe content shared by a Draft and a Recipe. It is what the
// model is asked to produce.
type Body struct {
	Title          string       `json:"title"`
	Description    string       `json:"description"`
	Servings       float64      `json:"servings"`
	ActiveMinutes  int          `json:"activeMinutes"`
	PassiveMinutes int          `json:"passiveMinutes"`
	Ingredients    []Ingredient `json:"ingredients"`
	Steps          []Step       `json:"steps"`
	Cookware       []Cookware   `json:"cookware"`
	Macros         Macros       `json:"macros"`
	Cuisine        *string      `json:"cuisine"`
	MealTypes      []string     `json:"mealTypes"`
	Cover          Cover        `json:"cover"`
}

// Ingredient is one line of a recipe: the name alone, a quantity in a closed
// unit set, how it is prepared, and the group it is listed under.
type Ingredient struct {
	ID          string   `json:"id"`
	Name        string   `json:"name"`
	Quantity    *float64 `json:"quantity"`
	Unit        *string  `json:"unit"`
	Preparation *string  `json:"preparation,omitempty"`
	Group       *string  `json:"group,omitempty"`
	Optional    bool     `json:"optional,omitempty"`
}

// Step names the Ingredients it first uses by id and carries no quantities or
// temperatures in its text (ADR 0004).
type Step struct {
	Text          string   `json:"text"`
	IngredientIDs []string `json:"ingredientIds"`
	TimerSeconds  *int     `json:"timerSeconds,omitempty"`
	TemperatureC  *float64 `json:"temperatureC,omitempty"`
}

// Cookware is one piece of cookware.
type Cookware struct {
	Name string `json:"name"`
}

// Macros are the Estimated Macros per serving. Source drives the "estimate"
// label in the app (H3).
type Macros struct {
	Source   string  `json:"source"`
	Calories float64 `json:"calories"`
	ProteinG float64 `json:"proteinG"`
	CarbsG   float64 `json:"carbsG"`
	FatG     float64 `json:"fatG"`
	FibreG   float64 `json:"fibreG"`
}

// MacrosSourceModelEstimate is the only macros source v1 writes.
const MacrosSourceModelEstimate = "model_estimate"

// Cover is the tagged visual of a recipe. A Draft always has kind "emoji".
type Cover struct {
	Kind      string  `json:"kind"`
	Emoji     string  `json:"emoji"`
	PhotoPath *string `json:"photoPath,omitempty"`
}

// CoverKindEmoji is the Cover kind every Draft carries.
const CoverKindEmoji = "emoji"

// Source is the provenance of a Draft or Recipe (H4). v1 writes only
// kind "generated".
type Source struct {
	Kind        string       `json:"kind"`
	URL         *string      `json:"url,omitempty"`
	Attribution *string      `json:"attribution,omitempty"`
	DerivedFrom *DerivedFrom `json:"derivedFrom,omitempty"`
}

// SourceKindGenerated is the only Source kind v1 writes.
const SourceKindGenerated = "generated"

// DerivedFrom points at the Recipe a Recipe was copied from. Never set in v1.
type DerivedFrom struct {
	RecipeID string `json:"recipeId"`
	Revision int    `json:"revision"`
}

File: app/lib/features/recipes/domain/recipe.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import '../../../core/firestore/rfc3339.dart';

part 'recipe.freezed.dart';
part 'recipe.g.dart';

// Dart side of the Go-to-Dart contract, schema/recipe.schema.json (ADR 0002,
// ADR 0004). Written by hand from the schema for now; the map's "Not yet
// specified" list holds generating both sides from the schema.
//
// Readers ignore unknown fields (json_serializable's default) and map unknown
// enum values to the fallback the schema names for each enum (H19).

/// Closed, metric-canonical unit set. `unknown` is the fallback for a value a
/// newer schema added: show the quantity with no unit and do not convert.
enum Unit { g, ml, tsp, tbsp, piece, clove, slice, sprig, bunch, pinch, unknown }

/// When a Recipe is eaten. An unknown value is dropped (see [MealTypesConverter]).
enum MealType { breakfast, lunch, dinner, snack, dessert, drink }

/// Drives the "estimate" label (H3). An unknown value is treated as an estimate.
enum MacroSource {
  @JsonValue('model_estimate')
  modelEstimate,
}

/// Provenance (H4). v1 only writes `generated`; an unknown kind renders the
/// recipe with no attribution.
enum SourceKind { generated, unknown }

/// Drops unknown meal types instead of failing the whole Recipe.
class MealTypesConverter
    implements JsonConverter<List<MealType>, List<dynamic>> {
  const MealTypesConverter();

  static final _byName = {for (final m in MealType.values) m.name: m};

  @override
  List<MealType> fromJson(List<dynamic> json) => [
    for (final v in json)
      if (v is String && _byName.containsKey(v)) _byName[v]!,
  ];

  @override
  List<dynamic> toJson(List<MealType> object) => [
    for (final m in object) m.name,
  ];
}

@freezed
abstract class Ingredient with _$Ingredient {
  const factory Ingredient({
    /// `i1`, `i2`, ... referenced by [Step.ingredientIds].
    required String id,

    /// The ingredient alone: "onion", not "1 large onion, finely diced".
    required String name,

    /// Null together with [unit] for "to taste".
    required double? quantity,
    @JsonKey(unknownEnumValue: Unit.unknown) required Unit? unit,
    String? preparation,
    String? group,
    @Default(false) bool optional,
  }) = _Ingredient;

  factory Ingredient.fromJson(Map<String, Object?> json) =>
      _$IngredientFromJson(json);
}

@freezed
abstract class Step with _$Step {
  const factory Step({
    /// Names ingredients without quantities and holds no temperatures; the
    /// app renders both beside the step in the viewer's Unit System.
    required String text,
    required List<String> ingredientIds,
    int? timerSeconds,
    double? temperatureC,
  }) = _Step;

  factory Step.fromJson(Map<String, Object?> json) => _$StepFromJson(json);
}

@freezed
abstract class Cookware with _$Cookware {
  const factory Cookware({required String name}) = _Cookware;

  factory Cookware.fromJson(Map<String, Object?> json) =>
      _$CookwareFromJson(json);
}

/// Estimated Macros, per serving.
@freezed
abstract class Macros with _$Macros {
  const factory Macros({
    @JsonKey(unknownEnumValue: MacroSource.modelEstimate)
    required MacroSource source,
    required double calories,
    required double proteinG,
    required double carbsG,
    required double fatG,
    required double fibreG,
  }) = _Macros;

  factory Macros.fromJson(Map<String, Object?> json) => _$MacrosFromJson(json);
}

/// The Cover (H13). The emoji is always kept, so removing a photo needs no
/// model call. An unknown `kind` falls back to the emoji.
@Freezed(unionKey: 'kind', fallbackUnion: 'emoji')
sealed class Cover with _$Cover {
  const factory Cover.emoji({required String emoji}) = EmojiCover;

  /// [photoPath] is a Cloud Storage path under the owner's prefix, never a
  /// download URL.
  const factory Cover.photo({
    required String emoji,
    required String photoPath,
  }) = PhotoCover;

  factory Cover.fromJson(Map<String, Object?> json) => _$CoverFromJson(json);
}

@freezed
abstract class DerivedFrom with _$DerivedFrom {
  const factory DerivedFrom({
    required String recipeId,
    required int revision,
  }) = _DerivedFrom;

  factory DerivedFrom.fromJson(Map<String, Object?> json) =>
      _$DerivedFromFromJson(json);
}

@freezed
abstract class Source with _$Source {
  const factory Source({
    @JsonKey(unknownEnumValue: SourceKind.unknown) required SourceKind kind,
    String? url,
    String? attribution,
    DerivedFrom? derivedFrom,
  }) = _Source;

  factory Source.fromJson(Map<String, Object?> json) => _$SourceFromJson(json);
}

/// What the Go service returns: a recipe the user has not saved. Holds nothing
/// about the Generation Request that produced it (H16).
@freezed
abstract class Draft with _$Draft {
  const factory Draft({
    /// UUIDv7 minted by the Go service; saving keeps it as the Recipe id.
    required String id,
    required int schemaVersion,
    required Source source,
    @Rfc3339DateTimeConverter() required DateTime createdAt,
    required String title,
    required String description,
    required double servings,
    required int activeMinutes,
    required int passiveMinutes,
    required List<Ingredient> ingredients,
    required List<Step> steps,
    @Default([]) List<Cookware> cookware,
    required Macros macros,
    String? cuisine,
    @MealTypesConverter() @Default([]) List<MealType> mealTypes,
    required Cover cover,
  }) = _Draft;

  factory Draft.fromJson(Map<String, Object?> json) => _$DraftFromJson(json);
}

/// A Rating (H9), owner-private, stored at `private.rating`.
@freezed
abstract class Rating with _$Rating {
  const factory Rating({
    required int stars,
    String? note,
    @Rfc3339DateTimeConverter() required DateTime ratedAt,
    required int recipeRevision,
  }) = _Rating;

  factory Rating.fromJson(Map<String, Object?> json) => _$RatingFromJson(json);
}

/// Owner-private data on the Recipe document (ADR 0007). A missing map or
/// member means unrated, in no Collection, never cooked.
@freezed
abstract class RecipePrivate with _$RecipePrivate {
  const factory RecipePrivate({
    Rating? rating,
    @Default([]) List<String> collectionIds,
    @Default(0) int cookCount,
    @Rfc3339DateTimeConverter() DateTime? lastCookedAt,
  }) = _RecipePrivate;

  factory RecipePrivate.fromJson(Map<String, Object?> json) =>
      _$RecipePrivateFromJson(json);
}

/// A saved Draft: the same body plus ownership and revision, stored at
/// `users/{uid}/recipes/{id}`. Renders from this document alone (H5).
@freezed
abstract class Recipe with _$Recipe {
  const factory Recipe({
    required String id,
    required int schemaVersion,
    required Source source,
    @Rfc3339DateTimeConverter() required DateTime createdAt,
    required String title,
    required String description,
    required double servings,
    required int activeMinutes,
    required int passiveMinutes,
    required List<Ingredient> ingredients,
    required List<Step> steps,
    @Default([]) List<Cookware> cookware,
    required Macros macros,
    String? cuisine,
    @MealTypesConverter() @Default([]) List<MealType> mealTypes,
    required Cover cover,
    required String ownerUid,
    required int revision,
    @Rfc3339DateTimeConverter() required DateTime savedAt,
    @Rfc3339DateTimeConverter() required DateTime updatedAt,
    RecipePrivate? private,
  }) = _Recipe;

  const Recipe._();

  factory Recipe.fromJson(Map<String, Object?> json) => _$RecipeFromJson(json);

  /// The Recipe that saving [draft] produces: the body unchanged, revision 1.
  /// The caller stores it with `savedAt` and `updatedAt` as server timestamps
  /// (see [firestoreDatePaths]).
  factory Recipe.fromDraft(
    Draft draft, {
    required String ownerUid,
    required DateTime savedAt,
  }) => Recipe(
    id: draft.id,
    schemaVersion: draft.schemaVersion,
    source: draft.source,
    createdAt: draft.createdAt,
    title: draft.title,
    description: draft.description,
    servings: draft.servings,
    activeMinutes: draft.activeMinutes,
    passiveMinutes: draft.passiveMinutes,
    ingredients: draft.ingredients,
    steps: draft.steps,
    cookware: draft.cookware,
    macros: draft.macros,
    cuisine: draft.cuisine,
    mealTypes: draft.mealTypes,
    cover: draft.cover,
    ownerUid: ownerUid,
    revision: 1,
    savedAt: savedAt,
    updatedAt: savedAt,
  );

  /// The date-time fields of a stored Recipe, for [FirestoreDates].
  static const firestoreDatePaths = [
    'createdAt',
    'savedAt',
    'updatedAt',
    'private.rating.ratedAt',
    'private.lastCookedAt',
  ];
}
